wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/remote/LsConfigurationExtractor.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.languageserver.remote;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.languageserver.registry.DocumentFilter;
+import org.eclipse.che.api.languageserver.registry.LanguageServerDescription;
+
+/**
+ * This class is responsible for language server description extraction out of server attributes
+ * map. It is expected that there will be specific attribute named <code>config</code> that will
+ * contain serialized json data that represents all language server configuration data. Structure of
+ * json corresponds to {@link LanguageServerDescription} class with all aggregated classes
+ */
+@Singleton
+class LsConfigurationExtractor {
+  private final JsonParser jsonParser;
+
+  @Inject
+  LsConfigurationExtractor(JsonParser jsonParser) {
+    this.jsonParser = jsonParser;
+  }
+
+  LanguageServerDescription extract(Map<String, String> attributes) {
+    String config = attributes.get("config");
```

Comment:
As we discussed before I would opt for a more FQN like name because the current one is not very descriptive.

Reply:
If we look at configuration fragment:
```json
{
   "environments":{
      "default":{
         "warnings":[

         ],
         "machines":{
            "typescript-ls-machine":{
               "env":{

               },
               "installers":[
                  "org.eclipse.che.exec",
                  "org.eclipse.che.terminal"
               ],
               "servers":{
                  "ls":{
                     "attributes":{
                        "internal":"true",
                        "type":"ls",
                        "config":"{\"id\":\"org.eclipse.che.plugin.web.typescript\", \"documentFilters\":[  {  \"languageId\":\"typescript\",  \"pathRegex\":\".*\\\\.(ts|tsx)\"} ]}"
                     },
                     "protocol":"tcp",
                     "port":"4417"
                  }
               },
               "volumes":{
                  "projects":{
                     "path":"/projects"
                  }
               },
               "attributes":{
                  "memoryLimitBytes":"1073741824"
               }
            }
         }
      }
   }
}
```
we see that `environments` have `machines` which have `servers` which have `attributes` so from this perspective it looks descriptive enough and the logical FQN is something like `default->typescript-ls-machine->ls->config`. On the other hand in the code in the decription it is mentioned: "class is responsible for language server description extraction out of server attributes map" so in underlined code line `attributes.get("config")` it should be rather obvious that keyword `cofig` is related to server's attributes while server is related to machin and so on. Again I'm wondering do you think that it is just `config` not descriptive or `internal` and `type` are not descriptive as well?

Reply:
I was concerned about attributes only since they do not have a fixed structure. And yes, I was thinking the same about `internal` and `type`.